### PR TITLE
#10861 : Added modf in numpy frontend 

### PR DESCRIPTION
**Pull request by RakshitKumar04**
Close https://github.com/unifyai/ivy/issues/12086

### CODE REVIEW

**Comment by RakshitKumar04** [comment]
Hey, @kurshakuz can you please guide me and help in passing these checks

**Comment by RakshitKumar04** [comment]
Hey, @kurshakuz I have updated the code, can please review my PR.

**Comment by RakshitKumar04** [comment]
Hey, @kurshakuz Please review my PR.

**Comment by RakshitKumar04** [comment]
Close #12086

**Comment by RakshitKumar04** [comment]
Hey @kurshakuz, Please review the PR.

**Review by kurshakuz — CHANGES_REQUESTED** [review]
Hey! Sorry for a delayed response! I have just checked your implementation and it seems to fail local tests. Can you please resolve below errors, check them locally and let me know once you are done? Thanks!
```
======================================================================== FAILURES =========================================================================
_____________________________________________ test_numpy_modf[cpu-ivy.functional.backends.numpy-False-False] ______________________________________________

arguments = (), kwargs = {}

    @handle_frontend_test(
        fn_tree="numpy.modf",
>       dtypes_and_x=helpers.dtype_and_values(
            available_dtypes=helpers.get_dtypes("float", "int"),
        ),
        where=np_frontend_helpers.where(),
    )
E   hypothesis.errors.InvalidArgument: test_numpy_modf() got an unexpected keyword argument 'dtypes_and_x', from `dtypes_and_x=dtype_and_values(available_dtypes=get_dtypes(kind='float', index='int'))` in @given

ivy_tests/test_ivy/test_frontends/test_numpy/test_mathematical_functions/test_trigonometric_functions.py:10: InvalidArgument
______________________________________________ test_numpy_modf[cpu-ivy.functional.backends.jax-False-False] _______________________________________________

arguments = (), kwargs = {}

    @handle_frontend_test(
        fn_tree="numpy.modf",
>       dtypes_and_x=helpers.dtype_and_values(
            available_dtypes=helpers.get_dtypes("float", "int"),
        ),
        where=np_frontend_helpers.where(),
    )
E   hypothesis.errors.InvalidArgument: test_numpy_modf() got an unexpected keyword argument 'dtypes_and_x', from `dtypes_and_x=dtype_and_values(available_dtypes=get_dtypes(kind='float', index='int'))` in @given

ivy_tests/test_ivy/test_frontends/test_numpy/test_mathematical_functions/test_trigonometric_functions.py:10: InvalidArgument
___________________________________________ test_numpy_modf[cpu-ivy.functional.backends.tensorflow-False-False] ___________________________________________

arguments = (), kwargs = {}

    @handle_frontend_test(
        fn_tree="numpy.modf",
>       dtypes_and_x=helpers.dtype_and_values(
            available_dtypes=helpers.get_dtypes("float", "int"),
        ),
        where=np_frontend_helpers.where(),
    )
E   hypothesis.errors.InvalidArgument: test_numpy_modf() got an unexpected keyword argument 'dtypes_and_x', from `dtypes_and_x=dtype_and_values(available_dtypes=get_dtypes(kind='float', index='int'))` in @given

ivy_tests/test_ivy/test_frontends/test_numpy/test_mathematical_functions/test_trigonometric_functions.py:10: InvalidArgument
_____________________________________________ test_numpy_modf[cpu-ivy.functional.backends.torch-False-False] ______________________________________________

arguments = (), kwargs = {}

    @handle_frontend_test(
        fn_tree="numpy.modf",
>       dtypes_and_x=helpers.dtype_and_values(
            available_dtypes=helpers.get_dtypes("float", "int"),
        ),
        where=np_frontend_helpers.where(),
    )
E   hypothesis.errors.InvalidArgument: test_numpy_modf() got an unexpected keyword argument 'dtypes_and_x', from `dtypes_and_x=dtype_and_values(available_dtypes=get_dtypes(kind='float', index='int'))` in @given

ivy_tests/test_ivy/test_frontends/test_numpy/test_mathematical_functions/test_trigonometric_functions.py:10: InvalidArgument
==================================================================== warnings summary =====================================================================
ivy/__init__.py:1164
  /workspaces/ivy/ivy/__init__.py:1164: UserWarning: jax module has been imported while ivy doesn't import it without setting a backend, ignore if that's intended
    warnings.warn(

ivy_tests/test_ivy/helpers/function_testing.py:69: 87 warnings
  /workspaces/ivy/ivy_tests/test_ivy/helpers/function_testing.py:69: DeprecationWarning: `np.bool8` is a deprecated alias for `np.bool_`.  (Deprecated NumPy 1.24)
    import ivy.compiler.compiler as ic

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
================================================================= short test summary info =================================================================
FAILED ivy_tests/test_ivy/test_frontends/test_numpy/test_mathematical_functions/test_trigonometric_functions.py::test_numpy_modf[cpu-ivy.functional.backends.numpy-False-False] - hypothesis.errors.InvalidArgument: test_numpy_modf() got an unexpected keyword argument 'dtypes_and_x', from `dtypes_and_x=dtype_and_values(available_...
FAILED ivy_tests/test_ivy/test_frontends/test_numpy/test_mathematical_functions/test_trigonometric_functions.py::test_numpy_modf[cpu-ivy.functional.backends.jax-False-False] - hypothesis.errors.InvalidArgument: test_numpy_modf() got an unexpected keyword argument 'dtypes_and_x', from `dtypes_and_x=dtype_and_values(available_...
FAILED ivy_tests/test_ivy/test_frontends/test_numpy/test_mathematical_functions/test_trigonometric_functions.py::test_numpy_modf[cpu-ivy.functional.backends.tensorflow-False-False] - hypothesis.errors.InvalidArgument: test_numpy_modf() got an unexpected keyword argument 'dtypes_and_x', from `dtypes_and_x=dtype_and_values(available_...
FAILED ivy_tests/test_ivy/test_frontends/test_numpy/test_mathematical_functions/test_trigonometric_functions.py::test_numpy_modf[cpu-ivy.functional.backends.torch-False-False] - hypothesis.errors.InvalidArgument: test_numpy_modf() got an unexpected keyword argument 'dtypes_and_x', from `dtypes_and_x=dtype_and_values(available_...
============================================================= 4 failed, 88 warnings in 1.78s ==============================================================
```

File: ivy/functional/frontends/numpy/mathematical_functions/trigonometric_functions.py [hunk 1]
```diff
@@ -11,6 +11,28 @@
 )
 
 
+@handle_numpy_out
+@handle_numpy_dtype
+@to_ivy_arrays_and_back
+@handle_numpy_casting
+@from_zero_dim_arrays_to_scalar
+def _modf(
```
Comment (kurshakuz) [inline ivy/functional/frontends/numpy/mathematical_functions/trigonometric_functions.py:19]: Can you please also clarify why do you need that underscore here? Thanks
```suggestion
def modf(
```

Comment (RakshitKumar04) [reply ivy/functional/frontends/numpy/mathematical_functions/trigonometric_functions.py]: I referred to the ivy frontends docs which state that "In NumPy, add() is categorised under mathematical_functions with a sub-category of arithmetic_operations as shown in the [numpy mathematical functions](https://numpy.org/doc/stable/reference/index.html) directory. It is important to note that add() is a universal function ([ufunc](https://numpy.org/doc/stable/reference/ufuncs.html)) in NumPy, thus the function is actually an object with instance methods like .at and .reduce, etc. We deal with this in the NumPy frontend by including a ufunc class and initialising it in the __init__ file" so thats why I used _modf.

https://lets-unify.ai/ivy/deep_dive/ivy_frontends.html

**Review by kurshakuz — CHANGES_REQUESTED** [review]
Hello @RakshitKumar04! I have tested your implementation and it is failing the tests. Please make sure it is correct and passes all of the tests, so that I can have a further look.

```
E               ivy.utils.exceptions.IvyException: Out argument must be an ivy.frontends.numpy.ndarray object
```

**Comment by RakshitKumar04** [comment]
Het, @kurshakuz I tried solving the error, please guide is it fails again

**Comment by RakshitKumar04** [comment]
Hey, @kurshakuz please review the PR.

**Comment by RakshitKumar04** [comment]
Hey, @kurshakuz please review my pr.

**Review by kurshakuz — CHANGES_REQUESTED** [review]
Please fix the following error and run the tests locally to see if your implementation works:
```
E               ivy.utils.exceptions.IvyException: Out argument must be an ivy.frontends.numpy.ndarray object
```

**Comment by ivy-seed** [comment]
This PR has been labelled as stale because it has been inactive for more than 7 days. If you would like to continue working on this PR, then please add another comment or this PR will be closed in 7 days.

**Comment by RakshitKumar04** [comment]
.

**Comment by RakshitKumar04** [comment]
Hey, @kurshakuz can you please guide me on resolving these errors. I ran the modf function on my local machine. Out of 5, 1st test case got passed and the other 4 showed the error listed below. 

>       return self._data.__array_ufunc__(*args, **kwargs)
**E       AttributeError: 'ArrayImpl' object has no attribute '__array_ufunc__'**
E       Falsifying example: test_numpy_modf(
E           dtype_and_x=(['float32'], [array(-1., dtype=float32)]),
E           fn_tree='ivy.functional.frontends.numpy.modf',
E           test_flags=FrontendFunctionTestFlags(
E               num_positional_args=1,
E               with_out=False,
E               inplace=False,
E               as_variable=[False],
E               native_arrays=[False],
E               generate_frontend_arrays=False,
E           ),
E           on_device='cpu',
E           frontend='numpy',
E       )

>     self.__getattribute__(name)
**E     AttributeError: 'tensorflow.python.framework.ops.EagerTensor' object has no attribute '__array_ufunc__'**
E     Falsifying example: test_numpy_modf(
E         dtype_and_x=(['float32'], [array(-1., dtype=float32)]),
E         fn_tree='ivy.functional.frontends.numpy.modf',
E         test_flags=FrontendFunctionTestFlags(
E             num_positional_args=1,
E             with_out=False,
E             inplace=False,
E             as_variable=[False],
E             native_arrays=[False],
E             generate_frontend_arrays=False,
E         ),
E         on_device='cpu',
E         frontend='numpy',
E     )

>       return self._data.__array_ufunc__(*args, **kwargs)
**E       AttributeError: 'Tensor' object has no attribute '__array_ufunc__'**
E       Falsifying example: test_numpy_modf(
E           dtype_and_x=(['float32'], [array(-1., dtype=float32)]),
E           fn_tree='ivy.functional.frontends.numpy.modf',
E           test_flags=FrontendFunctionTestFlags(
E               num_positional_args=1,
E               with_out=False,
E               inplace=False,
E               as_variable=[False],
E               native_arrays=[False],
E               generate_frontend_arrays=False,
E           ),
E           on_device='cpu',
E           frontend='numpy',
E       )

>       return self._data.__array_ufunc__(*args, **kwargs)
**E       TypeError: 'NoneType' object is not callable**
E       Falsifying example: test_numpy_modf(
E           dtype_and_x=(['float32'], [array(-1., dtype=float32)]),
E           fn_tree='ivy.functional.frontends.numpy.modf',
E           test_flags=FrontendFunctionTestFlags(
E               num_positional_args=1,
E               with_out=False,
E               inplace=False,
E               as_variable=[False],
E               native_arrays=[False],
E               generate_frontend_arrays=False,
E           ),
E           on_device='cpu',
E           frontend='numpy',
E       )

**Comment by kurshakuz** [comment]
@RakshitKumar04 I have not seen this error previously as well. I would suggest to address that issue in one of the Discord channels, maybe some has experience of dealing with that error. Can you do that?

**Comment by RakshitKumar04** [comment]
@kurshakuz Yes Sir, I will do that.

**Comment by RakshitKumar04** [comment]
Hey, @kurshakuz Sorry but I am unable to resolve the error in my implementation. Should I close the issue "Close as not planned"

**Comment by RakshitKumar04** [comment]
``` >       assert ret_dtype == ret_from_gt_dtype, (
            "the ground truth framework {} returned a {} datatype while "
            "the backend {} returned a {} datatype".format(
                ground_truth_backend,
                ret_from_gt_dtype,
                ivy.current_backend_str(),
                ret_dtype,
            )
        )
E       AssertionError: the ground truth framework numpy returned a float16 datatype while the backend numpy returned a uint8 datatype
E       Falsifying example: test_numpy_modf(
E           dtypes_values_casting=(['uint8'], [array(1, dtype=uint8)], 'no', None),
E           where=[array(False)], ```

Hey @kurshakuz how do I resolve this error?

**Comment by hirwa-nshuti** [comment]
Hey @RakshitKumar04  and @kurshakuz 

It seems like is the casting issue that is on our testing side 😁
I am fixing that and I think we can merge this PR.

**Pull request merged**